### PR TITLE
Add another example for how to use drand with single-owner objects

### PR DESCRIPTION
**Pull request by benr-ml**

### CODE REVIEW

**Comment by benr-ml** [comment]
@kchalkias no tests yet, just implemented the main logic. Please take a look and see if you have comments about the general design. Note that we need at least 2 epochs since users cannot check when an epoch began at the moment.

File: sui_programmability/examples/games/sources/drand_based_scratch_card.move [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    /// Create a new game with a given reward.
+    ///
+    /// The reward must be a positive balance, dividable by reward_factor. reward/reward_factor will be the ticket
+    /// price. base_drand_round is the current drand round.
+    public entry fun create(
+        reward: Coin<SUI>,
+        reward_factor: u64,
+        creator: address,
```
Comment (kchalkias) [inline sui_programmability/examples/games/sources/drand_based_scratch_card.move:80]: Awesome! yes this pattern!

File: sui_programmability/examples/games/sources/drand_based_scratch_card.move [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        let reward = Reward {
+            id: object::new(ctx),
+            game_id: object::id(&game),
+            balance: coin::into_balance(reward),
+        };
+        transfer::freeze_object(game);
+        transfer::share_object(reward);
+    }
+
+    /// Buy a ticket for a specific game, costing reward/RewardFactor SUI. Can be called only during the epoch in which
```
Comment (kchalkias) [inline sui_programmability/examples/games/sources/drand_based_scratch_card.move:104]: reward/RewardFactor either maintain lowercase `reward/reward_factor` or Reward with capital R `Reward/RewardFactor`

File: sui_programmability/examples/games/sources/drand_based_scratch_card.move [hunk 3]
```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// A basic game that depends on randomness from drand (chained mode).
+///
+/// Anyone can create a new game by depositing X*100 SUIs as a reward, and setting the current drand round as the base
+/// round. This creates two objects:
+/// - Game - an immutable object that includes all parameters to be used when buying tickets.
+/// - Reward - a shared object that holds the reward. It can be withdrawn by any winner ("first come, first served").
```
Comment (kchalkias) [inline sui_programmability/examples/games/sources/drand_based_scratch_card.move:9]: I'm happy to publish this game type, but ideally due to this FCFS rule we shouldn't recommend this pattern in practice for most real world applications. We need a warning message at the top of this file, if we keep this game.

Ideally, for the future we need a multi-player game with sortition and challenge period, to ensure one winner.

File: sui_programmability/examples/games/sources/drand_based_scratch_card.move [hunk 4]
```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// A basic game that depends on randomness from drand (chained mode).
+///
+/// Anyone can create a new game by depositing X*100 SUIs as a reward, and setting the current drand round as the base
```
Comment (kchalkias) [inline sui_programmability/examples/games/sources/drand_based_scratch_card.move:6]: **Let's make clear that the prize is fixed and independent on the num purchased tickets**. This is an important detail and the most common question (also the pattern on when this single write approach can be applied securely).

**Review by kchalkias — CHANGES_REQUESTED** [review]

File: sui_programmability/examples/games/sources/drand_based_scratch_card.move [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+///
+/// A user who wishes to play game G should:
+/// - Check if G.epoch is the current epoch, and that G.base_drand_round + 24h is in the future.
+///   See drand_based_lottery.move for how to calculate a round for a given point in time.
+/// - Check that the relevant reward is still non negative.
+/// - Call buy_ticket() with the right amount of SUI.
+/// - Wait until the relevant drand round's randomness is available, and can call evaluate().
+/// - If received an object Winner, claim the reward using take_reward().
+///
+/// Importnat: There is 1 reward per game, and there is no limit on the number of tickets that can be bought for a
```
Comment (kchalkias) [inline sui_programmability/examples/games/sources/drand_based_scratch_card.move:21]: Importa**nt** (typo)

File: sui_programmability/examples/games/sources/drand_based_scratch_card.move [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+/// a single game. *Only* the winner of a game who called take_reward first will receive a reward.
+/// One may extend this game and add another round in which winners can register their winner tickets, and then one of
```
Comment (kchalkias) [inline sui_programmability/examples/games/sources/drand_based_scratch_card.move:23]: Thanks!!

File: sui_programmability/examples/games/sources/drand_based_scratch_card.move [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    public fun get_game_base_epoch(game: &Game): u64 {
+        game.base_epoch
+    }
+
+    public fun end_of_game_round(round: u64): u64 {
+        // Since users do not know when an epoch has began, they can only check if the game depends on a round that is
+        // at least 24 hours from now. Since the creator does not know as well if its game is created in the beginning
+        // or the end of the epoch, we define the end of the game to be 24h + 24h from when it started, +1h to be on
+        // the safe side since epoch duration is not deterministic.
+        round + 2 * 60 * (24 + 25)
```
Comment (kchalkias) [inline sui_programmability/examples/games/sources/drand_based_scratch_card.move:191]: Indeed, acceptable assumption

**Review by kchalkias — APPROVED** [review]

**Pull request merged**